PEP 761: Deprecating PGP signatures for CPython artifacts (#4027)
python/peps: peps/pep-0761.rst

PGP's design requires the maintenance and protection of `long-lived private keys <https://words.filippo.io/giving-up-on-long-term-pgp/>`_ by trusted parties. PGP's `security and ergonomics have been criticized by security practitioners <https://www.latacora.com/blog/2019/07/16/the-pgp-problem/>`_ for many years now, with the biggest issue being that there were few alternatives for "artifact signing" being proposed or adopted.

Sigstore's design philosophy has focused on the ergonomics of signing and verifying and uses short-lived keys with strongly-bound human-readable identities via OpenID Connect. Sigstore has both development and adoption momentum, seeing adoption by PyPI, NPM, Homebrew, GitHub, among other ecosystems.

This PEP proposes to move CPython to using Sigstore exclusively for signing artifacts through a deprecation and eventual discontinuance of providing PGP signatures with new release managers.

Motivation
==========

CPython's releases are release-manager-centric, where a single person maintains multiple CPython releases from pre-release to end-of-life over the course of many years.

> Requiring release managers to maintain and protect PGP private keys for seven or more years is an unnecessary burden in the new age of ergonomic and ephemeral signing keys. Comparatively, Sigstore only requires release managers to click a button during the release process to OAuth sign-on to their identity provider. Maintaining the integrity of accounts on identity providers like GitHub is already an expectation of being a Python release manager or core developer, such as through multi-factor authentication and strong unique passwords.

Reviewer: Using language of PEP 13:
```suggestion
Requiring release managers to maintain and protect PGP private keys for seven or more years is an unnecessary burden in the new age of ergonomic and ephemeral signing keys. Comparatively, Sigstore only requires release managers to click a button during the release process to OAuth sign-on to their identity provider. Maintaining the integrity of accounts on identity providers like GitHub is already an expectation of being a Python release manager or core team member, such as through multi-factor authentication and strong unique passwords.
```

In general, the security model required for artifact signatures is being able to detect whether a given artifact is from the expected source and hasn't been modified, regardless of the security or integrity of the hosting service (in CPython's case: python.org/downloads).

`Sigstore's security model <https://docs.sigstore.dev/about/security/>`_ depends more on centralized infrastructure compared to PGP, such as the "public good" signature transparency log (Rekor), certificate authority and transparency log (Fulcio), and the security of OpenID Connect identity providers like Google and GitHub.

CPython's development already depends on the security of some of these services and the others are better resourced than any individual release manager to provide long-term public key management.

How to Teach This
=================

CPython `already documents <https://www.python.org/downloads/metadata/sigstore/>`_ how to verify artifacts using Sigstore based on the pre-published identities of release managers. Documentation will be updated to indicate the deprecation and future expectations of PGP signatures.

> Verifying signatures of CPython artifacts isn't something we should expect from new Python users. Instead, Sigstore is more likely to be a part of a downstream integrators' build pipeline such as a Linux distro, Homebrew, pyenv, or others that programmatically fetch and build CPython from source.

Reviewer: ```suggestion
Verifying signatures of CPython artifacts isn't something we should expect from new Python users. Instead, Sigstore is more likely to be a part of a downstream integrator's build pipeline such as a Linux distro, Homebrew, pyenv, or others that programmatically fetch and build CPython from source.
```

Rejected Ideas
==============

Continue publishing PGP signatures indefinitely
-----------------------------------------------

> Being a release manager is already a difficult, time-consuming, and long-term commitment that is typically done on a volunteer basis. Thus we see removal of PGP key management duties as a step towards reducing burnout and stress of future release managers and improving the sustainability of CPython.

Reviewer: ❤️

Appendix
========

Support for offline verification
--------------------------------

> During the `pre-PEP discussion <https://discuss.python.org/t/pre-pep-discussion-stop-providing-gpg-signatures-for-cpython-artifacts/65058>`_, there was a question of whether offline verification was supported by Sigstore. Using a Sigstore bundle (`.sigstore`) file, `Sigstore clients support verifying the artifact completely offline <https://discuss.python.org/t/pre-pep-discussion-stop-providing-gpg-signatures-for-cpython-artifacts/65058/9>`_. When in offline mode, Sigstore can't verify whether a signature has been revoked. This is a similar restriction to PGP key revocations not being detectable during offline verification.

Reviewer: ```suggestion
During the `pre-PEP discussion <https://discuss.python.org/t/pre-pep-discussion-stop-providing-gpg-signatures-for-cpython-artifacts/65058>`_, there was a question of whether offline verification was supported by Sigstore. Using a Sigstore bundle (:file:`.sigstore`) file, `Sigstore clients support verifying the artifact completely offline <https://discuss.python.org/t/pre-pep-discussion-stop-providing-gpg-signatures-for-cpython-artifacts/65058/9>`_. When in offline mode, Sigstore can't verify whether a signature has been revoked. This is a similar restriction to PGP key revocations not being detectable during offline verification.
```

on a "release manager boundary", a new Python release could be a potential
boundary.

Because the primary motivation for deprecating PGP is ergonomics, deciding
to drop PGP for one release while a release manager still has obligations to
provide PGP signatures for other releases for multiple years isn't much
savings of effort.

A new release manager also represents a new PGP public key that downstream
verifiers need to adopt. By choosing to make the change during this period,
this minimizes the breakage to a place in downstream maintenance where a
change will already be necessary.

Gordian knot of signing methods and verifiers
---------------------------------------------

CPython providing both PGP and Sigstore signatures concurrently creates a
"`Gordian knot <https://en.wikipedia.org/wiki/Gordian_Knot>`_" where
verifiers are disincentivized to migrate to a new signature method due to the
*continued and expected availability* of an existing signature method, thus
propagating the *apparent demand* for maintaining the existing signature
method.

This situation slows down the adoption of Sigstore downstream vertically

> between ecosystems and horizontally across ecosystems by not creating a "need"

Author: I'll come up with a clearer phrasing, was meant to distinguish between inter-ecosystem adoption and intra-ecosystem adoption :)

> between ecosystems and horizontally across ecosystems by not creating a "need"

Author: https://github.com/python/peps/pull/4027/commits/1397ef9a80fc7e76c69de589a97044df96ae158f

propagating the *apparent demand* for maintaining the existing signature
method.

> This situation slows down the adoption new signature methods like Sigstore for

Reviewer 2: ```suggestion
This situation slows down the adoption of new signature methods like Sigstore for
```
